build-system/tasks/runtime-test/helpers.js

```diff
@@ -249,9 +289,402 @@ function isLargeRefactor() {
+}
+
+/**
+ * Runs tests in batches.
+ *
+ * Splits stable and beta browsers to separate batches. Test failures in any
+ * of the stable browsers will return an exit code of 1, whereas test failures
+ * in any of the beta browsers will only print error messages, but will return
+ * an exit code of 0.
+ *
+ * @param {Object} config karma config
+ * @return {number} exitCode
+ */
+async function runTestInBatches(config) {
+  const browsers = {stable: [], beta: []};
+  for (const browserId of config.browsers) {
+    browsers[
+      browserId.toLowerCase().endsWith('_beta') ? 'beta' : 'stable'
+    ].push(browserId);
+  }
+  if (browsers.stable.length) {
+    const allBatchesExitCodes = await runTestInBatchesWithBrowsers(
+      'stable',
+      browsers.stable,
+      config
+    );
+    if (allBatchesExitCodes) {
+      log(
+        yellow('Some tests have failed on'),
+        cyan('stable'),
+        yellow('browsers, so skipping running them on'),
+        cyan('beta'),
+        yellow('browsers.')
+      );
+      return allBatchesExitCodes;
+    }
+  }
+
+  if (browsers.beta.length) {
+    const allBatchesExitCodes = await runTestInBatchesWithBrowsers(
+      'beta',
+      browsers.beta
+    );
+    if (allBatchesExitCodes) {
+      log(
+        yellow('Some tests have failed on'),
+        cyan('beta'),
+        yellow('browsers.')
+      );
+      log(
+        yellow('This is not currently a fatal error, but will become an'),
+        yellow('error once the beta browsers are released as next stable'),
+        yellow('version!')
+      );
+    }
+  }
+
+  return 0;
+}
+
+/**
+ * Runs tests in named batch(es), with the specified browsers.
+ *
+ * @param {string} batchName a human readable name for the batch.
+ * @param {!Array{string}} browsers list of SauceLabs browsers as
+ *     customLaunchers IDs.
+ * @param {Object} config karma config
+ * @return {number} processExitCode
+ */
+async function runTestInBatchesWithBrowsers(batchName, browsers, config) {
+  let batch = 1;
+  let startIndex = 0;
+  let endIndex = BATCHSIZE;
+  const batchExitCodes = [];
+
+  log(
+    green('Running tests on'),
+    cyan(browsers.length),
+    green('Sauce Labs'),
+    cyan(batchName),
+    green('browser(s)...')
+  );
+  while (startIndex < endIndex) {
+    const configBatch = Object.assign({}, config);
+    configBatch.browsers = browsers.slice(startIndex, endIndex);
+    log(
+      green('Batch'),
+      cyan(`#${batch}`) + green(': Running tests on'),
+      cyan(configBatch.browsers.length),
+      green('Sauce Labs browser(s)...')
+    );
+    batchExitCodes.push(await createKarmaServer(configBatch));
+    startIndex = batch * BATCHSIZE;
+    batch++;
+    endIndex = Math.min(batch * BATCHSIZE, browsers.length);
+  }
+
+  return batchExitCodes.every(exitCode => exitCode == 0) ? 0 : 1;
+}
+
+async function createKarmaServer(config) {
+  let resolver;
+  const deferred = new Promise(resolverIn => {
+    resolver = resolverIn;
+  });
```

Comment:
This `resolver` / `deferred` pattern was necessary before `async` and `await` were available. Let's leave this as it is here, and try to remove in a separate PR.

Reply:
These aren't necessarily related. The deferred promise allows a separate thread to resolve a promise on the current thread (specifically, `Server` here is responsible for resolving the promise that this function returns) - I don't think `async/await` can help simplify the code here